pkg/resources/provisioning.cattle.io/v1/cluster/mutator_test.go

```diff
@@ -617,3 +621,275 @@ func TestAdmitPreserveUnknownFields(t *testing.T) {
+			request: &admission.Request{AdmissionRequest: admissionv1.AdmissionRequest{Operation: admissionv1.Update}},
+			cluster: &v1.Cluster{
+				ObjectMeta: metav1.ObjectMeta{Annotations: map[string]string{"provisioning.cattle.io/allow-dynamic-schema-drop": "false"}},
+				Spec: v1.ClusterSpec{
+					RKEConfig: &v1.RKEConfig{
+						MachinePools: []v1.RKEMachinePool{
+							{
+								Name: "a",
+							},
+						},
+					},
+				},
+			},
+			oldCluster: &v1.Cluster{
+				Spec: v1.ClusterSpec{
+					RKEConfig: &v1.RKEConfig{
+						MachinePools: []v1.RKEMachinePool{
+							{
+								Name:              "a",
+								DynamicSchemaSpec: "a",
+							},
+						},
+					},
+				},
+			},
+			expected: []v1.RKEMachinePool{
+				{
+					Name:              "a",
+					DynamicSchemaSpec: "a",
+				},
+			},
+		},
+		{
+			name:    "schema present on old cluster and new cluster with true annotation",
+			request: &admission.Request{AdmissionRequest: admissionv1.AdmissionRequest{Operation: admissionv1.Update}},
+			cluster: &v1.Cluster{
+				ObjectMeta: metav1.ObjectMeta{Annotations: map[string]string{"provisioning.cattle.io/allow-dynamic-schema-drop": "true"}},
+				Spec: v1.ClusterSpec{
+					RKEConfig: &v1.RKEConfig{
+						MachinePools: []v1.RKEMachinePool{
+							{
+								Name:              "a",
+								DynamicSchemaSpec: "a",
+							},
+						},
+					},
+				},
+			},
+			oldCluster: &v1.Cluster{
+				Spec: v1.ClusterSpec{
+					RKEConfig: &v1.RKEConfig{
+						MachinePools: []v1.RKEMachinePool{
+							{
+								Name:              "a",
+								DynamicSchemaSpec: "a",
+							},
+						},
+					},
+				},
+			},
+			expected: []v1.RKEMachinePool{
+				{
+					Name:              "a",
+					DynamicSchemaSpec: "a",
+				},
+			},
+		},
+		{
+			name:    "schema present on old cluster but not new cluster with true annotation",
+			request: &admission.Request{AdmissionRequest: admissionv1.AdmissionRequest{Operation: admissionv1.Update}},
+			cluster: &v1.Cluster{
+				ObjectMeta: metav1.ObjectMeta{Annotations: map[string]string{"provisioning.cattle.io/allow-dynamic-schema-drop": "true"}},
+				Spec: v1.ClusterSpec{
+					RKEConfig: &v1.RKEConfig{
+						MachinePools: []v1.RKEMachinePool{
+							{
+								Name: "a",
+							},
+						},
+					},
+				},
+			},
+			oldCluster: &v1.Cluster{
+				Spec: v1.ClusterSpec{
+					RKEConfig: &v1.RKEConfig{
+						MachinePools: []v1.RKEMachinePool{
+							{
+								Name:              "a",
+								DynamicSchemaSpec: "a",
+							},
+						},
+					},
+				},
+			},
+			expected: []v1.RKEMachinePool{
+				{
+					Name: "a",
+				},
+			},
+		},
+		{
+			name:    "new machine pool without schema",
+			request: &admission.Request{AdmissionRequest: admissionv1.AdmissionRequest{Operation: admissionv1.Update}},
+			cluster: &v1.Cluster{
+				ObjectMeta: metav1.ObjectMeta{Annotations: map[string]string{"provisioning.cattle.io/allow-dynamic-schema-drop": "true"}},
```

Comment:
Since this has the annotation allowing drop, it never gets to the statement that allows a bypass on the new machine pool. I would recommend removing the annotation so that this case is also covered.